### PR TITLE
Add flags to allow interactive debugging #715

### PR DESCRIPTION
**Pull request by fabioDMFerreira**
* Disable accel;

* Append noaslr flag;

* Check program is statically linked and has debugging symbols.

### CODE REVIEW

**Comment by eyberg** [comment]
the code here is fine, however, the syscall tracing needs to be separated as it vomits output, espc in larger applications whereas this functionality is more of an interactive step-by-step/breakpoint/watchpoint/etc..

I guess I'm open to suggestion on what we should do here but I think the syscall tracing should be put into a diff flag something and this turn into the actual 'debug' flag - but they don't belong together for sure

if anyone from @nanovms/engineering has comments feel free to chime in

**Comment by eyberg** [comment]
also - maybe it's worth popping some instructions to the end-user or a link to https://nanovms.gitbook.io/ops/debugging

**Comment by francescolavra** [comment]
I agree that debug and tracing should be two separate options.
I guess we could use --trace (with no shorthand synonym, because -t is already used to specify the target cloud platform) to enable kernel tracing, and -d / --debug to enable interactive debugging.
As a side note, I may be missing something but I don't see why acceleration would need to be disabled in order to allow interactive debugging. Maybe @eyberg can shed some light on why forcing accel off is in the checklist at https://github.com/nanovms/ops/issues/715?

**Comment by wjhun** [comment]
> As a side note, I may be missing something but I don't see why acceleration would need to be disabled in order to allow interactive debugging. Maybe @eyberg can shed some light on why forcing accel off is in the checklist at #715?

AFAIK, debugging via the qemu gdb stub only works when qemu is running with TCG.

**Comment by francescolavra** [comment]
I tested the gdb stub with KVM enabled, and it sort of works, but with (at least) one major inconvenience: it's not possible to insert breakpoints in the kernel until memory paging has been enabled. I believe this is due to the fact that with KVM enabled, GDB cannot use software breakpoints and must use hardware breakpoints instead.
Without acceleration this problem does not appear. So it makes perfect sense for Ops to force accel off when the debug flag is set.

**Comment by eyberg** [comment]
@fabioDMFerreira - so if it's possible to have the syscall trace be '--trace' (and keep -t for cloud selection for now), and then convert '-d' to actual debugger let's do that and then we'll merge this

File: cmd/run.go [hunk 1]
```diff
@@ -141,10 +151,33 @@ func runCommandHandler(cmd *cobra.Command, args []string) {
...
+		if api.IsDynamicLinked(elfFile) {
+			log.Fatalf("Program %s must be linked statically", c.ProgramPath)
+		}
+
+		if !api.HasDebuggingSymbols(elfFile) {
+			log.Fatalf("Program %s must be compiled with debugging symbols", c.ProgramPath)
+		}
+	}
+
+	if trace || debug {
```
Comment (eyberg) [inline cmd/run.go:177]: this is just for debug not for trace

**Comment by eyberg** [comment]
this is failing to build on osx cause the folowing is only defined in ldd_linux - the underlying lib might only build in linux so i'm ok stubbing this out for now in ldd_darwin but there's no reason why they shouldn't be able to work under osx

**Comment by eyberg** [comment]
excellent - thanks!

**Pull request merged**